Master api options

vid-n3t opened the pull request:
Hi Peter; 

Thanks for building the dao-datascraper.
I continued to develop on this for my use today.

This pull request includes the previous request along with the remaining API options that etherscan offers.

vid-n3t commented on main.js:
```diff
@@ -23,16 +30,58 @@ const formattedDate = {
+const outputPath = `output/${outputFileNamePrepend}${formattedDate.year}_${formattedDate.month}_${formattedDate.day}`;
 // csv field headers
-const fields = ['blockNumber', 'timeStamp', 'hash', 'nonce', 'blockHash', 'transactionIndex', 'from', 'to', 'value', 'gas', 'gasPrice', 'input', 'contractAddress', 'cumulativeGasUsed', 'gasUsed', 'confirmations'];
+var fields = ['blockNumber', 'timeStamp', 'hash', 'nonce', 'blockHash', 'transactionIndex', 'from', 'to', 'value', 'gas', 'gasPrice', 'isError', 'txreceipt_status', 'input', 'contractAddress', 'cumulativeGasUsed', 'gasUsed', 'confirmations'];
```
Changed this const to a var as it's overridden below in the case of internal transactions (txlistinternal). Also added two new columns available from the etherscan json output.

vid-n3t commented on main.js:
```diff
@@ -23,16 +30,58 @@ const formattedDate = {
 
 process.stdout.write('DAO-datascraper initialized!\n');
 
+var apiURL = `https://api.etherscan.io/api?module=account&address=${address}`;
```
The apiURL is built from the specified options in settings.js. if none are specified then defaults are set for 'sort=asc' and 'action=txlist'

vid-n3t commented on settings.js:
```diff
@@ -1,10 +1,11 @@
-    , sortOrder: 'asc' //'desc|asc'
-    , apikey: ''; //YourApiKeyToken //blank for none
-};
+   address: '0xbb9bc244d798123fde783fcc1c72d3bb8c189413', //REQUIRED //TheDao See: https://etherscan.io/address/0xbb9bc244d798123fde783fcc1c72d3bb8c189413
```
Finished adding fields; moved the commas to the end of each entry.

PeterChauYEG commented:
Hey there,

It these look good! As you can tell, I have worked on these on a while. I'd be happy to merge these. 

Can I ask your use-case? Perhaps I can help you get this up to date.

vid-n3t commented:
HI Peter; 

Sure;

I’m using your dao-datascraper for research. 
I wanted to capture all the tx from a new token and save that off in to CSV. 

Initially, I was working on doing that with the web3 api; 
At the time I didn’t have an elegant way to capture and parse the JSON into CSV. 
Ultimately I hacked together a one-off but I knew I wanted a nice reusable solution. So I was happy to find your project, which works great. 
Very well done.
Thanks for sharing!

The initial additions I made were to accommodate the number transactions for a specific token (19k+), since it exceeded etherscan’s 10k limit.

Then I thought I’d just accommodate all the API options I could find. 
Then in testing, I noticed the CSV output was different for internal tx, I changed that csv list from a const to a var, so we can have unique columns for each type of query.

This new token forked recently and I’m going to take the combined output of the dao-datascraper (all 19k+ tx) to generate a report on the balance of each account at the block of each transaction, as well as a report for each unique account holder at the fork and a few points afterwards.

Thus far, I’ve combined the two outputs into one spreadsheet now and removed duplicates. 
Next I’ll make the array to loop through for the big report then I’ll pull unique accounts and run those through a report to get the bal at the fork and beyond; though that may be redundant if the data from the first report is usable in excel.

I’ve written a bash wrapper and node script to do the eth_call for each account and populate a new csv file. I suspect that will be running all night!


In the future I suspect I’ll want a way to do one-offs; 
e.g. take a single account and capture all it’s transactions for a specific token and then get the token balance at each transaction block. 
Like you see in most wallets. 
Also I’d like to capture the amount of the token that is being transferred in each transaction. You see than on Etherscan now (Token Transfer: … )

I haven’t gotten quite that deep into the eth_call options to sort those bits out. 

But, implementing that in a React-Native wallet (currently under development) will have to happen soon!

Let me know if you’re looking for work!

Vid

> On Mar 26, 2018, at 4:17 PM, Peter C <notifications@github.com> wrote:
> 
> Merged #2 <https://github.com/PeterChauYEG/dao-datascraper/pull/2>.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/PeterChauYEG/dao-datascraper/pull/2#event-1542053307>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ADwZutKiBDK7OinBl20BXAEqaugoYKUFks5tiXcYgaJpZM4S8AFt>.
>
